### DISCUSSION TITLE
PVE vm Sriov network interfaces problem

### DISCUSSION THREAD

**Discussion by Geilberton** [kkkgo/PaoPaoGateWay#28] [问和答] [answered]
docker下定制的iso，打包完发现无法识别直通的pcie网卡，该网卡是sriov出来的vf（其他vm正常使用vf）。启动后一直显示ppgateway waitting for eth0。只有添加虚拟网卡才可以正常连接， 请问是否有可以修改识别直通网卡的相关设置？该iso是在centos7下搭建的docker定制的。
辛苦~
感谢~

**Comment by kkkgo** [comment]
镜像因为是虚拟机专用默认仅包含虚拟网卡驱动。
另外我记得sriov的vf驱动跟直通驱动还不太一样，你尝试过使用任意openwrt发行版的VM可以驱动你的网卡吗？如果可以的话，有办法添加驱动。

**Comment by Geilberton** [comment]
用p3r大佬的action自编的lienol 2203没有任何问题，可以正确识别并且运行无Bug，其他版本的话没试过呢~
意思是需要和在windows添加独立的trust吗？还是只是注入驱动就可以？

**Comment by kkkgo** [comment]
你在可以运行驱动的openwrt上，安装pciutils包`opkg install pciutils `
然后执行`lspci|grep net`，看看网卡相关的驱动显示的是什么

**Comment by Geilberton** [comment]
06:10.0 Ethernet controller: Intel Corporation I350 Ethernet Controller Virtual Function (rev 01)
4口的I350噢，我的物理网卡，之前在pve上开启sriov时我也试过，都是这个驱动。

**Comment by kkkgo** [comment]
我尝试一下，再给你定制方法

**Comment by Geilberton** [comment]
非常感谢~
辛苦大佬~
好人平安~
如果实在不行我就用回虚拟网卡吧，毕竟如果需要单独注入一个驱动包太辛苦大佬了~

**Comment by kkkgo** [comment, marked as answer]
把定制的docker镜像切换成`fullmod`版本：
```shell
docker pull sliamb/ppgwiso:fullmod
docker run --rm -v .:/data sliamb/ppgwiso:fullmod
```

**Comment by kkkgo** [comment]
`sliamb/ppgwiso:fullmod`附带了所有可能支持的网卡驱动和相关模块，生成的镜像会大20M左右，可适当增加运行内存。

**Comment by Geilberton** [comment]
非常感谢大佬~感激不尽~
iso由62增大到82，不过是可以接受的，毕竟只是网关而不是dns。ram占用在210+左右，分配256也够用了，后期不够在加到384。我有空在对照下虚拟驱动版本的iso把其他不需要清理下，留下单独的igb驱动就好了。我本来想自己去找一个直接扔进moduels里面的。但是找了下好象都是.c的驱动，并没有直接的.ko。所以就放弃了~
另外想请教下大佬，我用的centos7搭建的ppdns，命令如下： 
`
# 启用eth0网卡混杂模式
ip link set enp1s0 promisc on
# 创建macvlan网络
docker network create -d macvlan --subnet=10.10.10.0/24 --gateway=10.10.10.13 -o parent=enp1s0 macvlan1
#拉取最新的docker镜像
docker pull sliamb/paopaodns:latest
# 运行容器并指定IP
docker run -d \
--name paopaodns \
-v /home/mydata:/data \
-e TZ=Asia/Shanghai \
-e DNS_SERVERNAME=PaoPaoDNS,blog.03k.org \
-e DNSPORT=53 \
-e UPDATE=weekly \
-e USE_HOSTS=no \
-e CNAUTO=yes \
-e CNFALL=yes \
-e CN_TRACKER=yes \
-e IPV6=no \
-e SOCKS5=10.10.10.13:1080 \
-e CUSTOM_FORWARD=10.10.10.13:53  \
-e  AUTO_FORWARD=yes   \
-e  AUTO_FORWARD_CHECK=yes \
-e USE_MARK_DATA=yes \
-e HTTP_FILE=yes \
--restart always \
--network macvlan1 --ip 10.10.10.18 \
-p 53:53/tcp -p 53:53/udp \
-p 5304:5304/tcp -p 5304:5304/udp \
-p 7889:7889/tcp \
sliamb/paopaodns
`
搭建完成后，虽然显示跑起来了，但是依然不可达？ping dns 18的时候并不可达（不是宿主机，主路由没有看到相关arp显示，宿主机的ip是可以ping通的:10.10.10.17）。docker status 是正常运行的，请问是我配置有问题么？
请原谅我的知识匮乏，我只是对着教程来做，并没有太多相关的知识。谢谢~

**Comment by kkkgo** [comment]
PVE可以参考这个，配置写的很详细清晰：https://github.com/kkkgo/PaoPaoGateWay/discussions/23  
带驱动的定制镜像跑起来了吗？

**Comment by kkkgo** [comment]
如果网卡直通能正常工作的话我更新一下文档，有类似需求切换`fullmod`版本即可。

**Comment by Geilberton** [comment]
是的，正常跑起来了，网卡可以正确识别了~
后期我再自己去掉那些用不上的就可以缩减体积和ram了。
我需求是只让fakeip dns分流到ppdns解析就行了，所以gateway才设置的ppgw，因为我主路由是ros，有设置了ip分流。ppdns建立时不想经过二次docker nat，所以想建立macvlan独立桥接这样，实在不行我再试试装个ui管理吧，或者直接用原docker nat的网络算了。
总之，非常感谢~
辛苦大佬~

**Reply by kkkgo** [reply]
不太理解你描述的拓扑，fakeip已经是fake的，为啥要分流到ppdns，有IP分流的话，fakeip的作用是在哪个流程？
fakeip其实是算是为了解决IP分流(real ip)的缺点的东西，这样就不用维护更新IP库、IP库不准确、BT下载和P2P应用打洞等问题，维护域名也比维护IP段灵活。

**Reply by Geilberton** [reply]
打着打着就迷胡了，原义是只需要ros cnip分流过后的ip才由静态路由表也就是fakeip直接返回给ppgw，其他的全部由主路由过去，实现真正的 旁 路由，而不是网关互指这样。
目的就是和大佬你说的一样，不然也不用奔着大佬的ppgw来了，有看到有些大佬写的基于linux的clash，但是都没有看到有fake ip而且也没有后续进一步的进阶说明。
原来用的op里面的opclash太烦人了，fakeip缓存好象也有问题，手动清理也不起作用（无论是否持久化），经常两三天不重启就不行，而且有时某个依赖或者程序跑不起来一直重启，装了个watchcat自动重启才有所改善。索性就直接放弃了重新找方案。

**Reply by kkkgo** [reply]
你用PaoPaoDNS的话，并不需要在ROS上设置任何CN IP分流啊，一切都基于域名来分流（除了一些IP直连的应用），PaoPaoDNS的`AUTO_FORWARD`功能就是你所描述的非CN IP自动转发到其他DNS解析，配合PPGW的话转发过去就是FAKE IP，一条静态路由就完成了非CN域名的流量走PPGW了。
FAKEIP缓存的话，可以参考docker定制的可选前置嗅探一节，可以改善这种情况。

**Reply by kkkgo** [reply]
参考FAKEIP网关工作原理： https://blog.03k.org/post/paopaogateway.html

**Reply by Geilberton** [reply]
是的，我搭建好后就发现了，因为我局域网里面只是分配了部份的设备海淘，dhcp下的地址池设备都是不走ppdns，所以我就把原来的ip分流给禁用了.
然后在ppdns前配置了一个前置的dns用于转发不同局域网的cidr，因为之前搭建的adgh还留着没有删掉就拿来继续用了，只有海淘的才转发到ppdns，现在ros除了tg和nf也只有一条fakeip路由表转发到ppgw，干净又整洁，docker定制时已经添加了嗅探了，有细看说明书的，非常感谢大佬。
现在ppgw配和ppdns挺好，dns leak也和之前一样基本没什么问题。
非常感谢author大佬~好人健康平安~

**Comment by kkkgo** [comment]
你可以参考上面那个PVE的安装分享，如果macvlan比较麻烦的话直接用host网络模式也可以，host网络就是直接本机网络。跑docker不一定是centos，alpine作为docker的虚拟机体积最小最简单。

**Reply by kkkgo** [reply]
https://www.cnblogs.com/freeaihub/p/13197292.html#!comments
